### PR TITLE
Webapp interventions

### PR DESCRIPTION
**Pull request by cliffckerr**
Implements functional interventions for the webapp. Not extensively tested, but seem to work. This PR builds off of (and includes) PR https://github.com/InstituteforDiseaseModeling/covasim/pull/169.

This version is currently live at https://covid.idmod.org/sim

Closes https://github.com/InstituteforDiseaseModeling/covasim/issues/75 https://github.com/InstituteforDiseaseModeling/covasim/issues/134 https://github.com/InstituteforDiseaseModeling/covasim/issues/177

### CODE REVIEW

**Comment by lgeorge-idm** [comment]
@cliffckerr , there's no point to having a separate PR in #169 if they're combined into this one.  I would suggest you only include the model changes or close  #169 .

**Comment by ckerr-IDM** [comment]
@lgeorge-idm sure -- i did that in case you wanted to review (slightly) smaller PRs, but i'll close the other one

File: covasim/webapp/cova_app.py [hunk 1]
```diff
@@ -287,6 +291,8 @@ def run_sim(sim_pars=None, epi_pars=None, intervention_pars=None, datafile=None,
     # Create the sim and update the parameters
     try:
         sim = cv.Sim(pars=web_pars,datafile=datafile)
+        print('FDLKJDFLKDFLKJDLKJFDLKFDJLKFJKLDKLFDJLK')
```
Comment (mfisher-idmod) [inline covasim/webapp/cova_app.py:294]: are you sure you want this?

Comment (ckerr-IDM) [reply covasim/webapp/cova_app.py:294]: Goodness no, good catch, thank you!

File: covasim/webapp/index.html [hunk 2]
```diff
@@ -36,292 +36,300 @@
...
-    <b-sidebar id="parameters" :visible="panel_open" no-header bg-variant="white" :width="panel_width+'vw'" shadow>
+    <b-sidebar id="parameters" :visible="panel_open" no-header class="sidebar" :width="panel_width+'vw'" shadow>
       <div class="px-3">
         <header class="d-flex justify-content-between">
-          <h4 class="my-2">Campaign Builder</h4>
-          <button class="close" @click="close_panel" aria-label="Close">
-            <span aria-hidden="true">&times;</span>
+          <h4 class="my-2">Simulation configuration</h4>
+          <button class="close" id="hidepanel" @click="close_panel" aria-label="Left">
+            <span aria-hidden="true">🡄</span>
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:67]: Use an icon class or the hexidecimal code for this.

File: covasim/webapp/index.html [hunk 3]
```diff
@@ -36,292 +36,300 @@
...
+  </div>
   </div>
 
   <footer class="row no-gutter">
     <b-navbar type="dark" variant="dark">
-      <span class="navbar-text text-white">
-        &copy; {{ app.copyright_year }} <a :href="app.org_url" target="_blank">Institute for Disease Modeling</a>
-      </span>
+      <a :href="app.org_url" target="_blank">
+        <img src="assets/idm-logo.png" height="30px"/>
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:328]: Don't set height directly on an image. Use bootstrap responsive image classes.

**Comment by cliffckerr** [comment]
Thanks @mfisher-idmod and @lgeorge-idm , reviewing now

File: covasim/webapp/cova_app.py (hunk 1 above)
Comment (cliffckerr) [reply covasim/webapp/cova_app.py:294]: Fixed

File: covasim/webapp/cova_app.py [hunk 4]
```diff
@@ -190,11 +142,12 @@ def upload_file(file):
         fd.close()
     return path
 
+
 @app.register_RPC()
-def get_gantt(intervention_pars=None, intervention_config=None):
+def get_gantt(int_pars=None, intervention_config=None):
```
[the remark this answers is not in the document]
Comment (cliffckerr) [reply covasim/webapp/cova_app.py:147]: I had a look if I missed some in the refactor but I'm not seeing anything. I changed the name to `int_pars` to be consistent with `sim_pars` and `epi_pars`.

File: covasim/webapp/cova_app.py [hunk 5]
```diff
@@ -24,29 +26,34 @@
...
 def log_err(message, ex):
     ''' Compile error messages to send to the frontend '''
     tex = traceback.TracebackException.from_exception(ex)
-    print(f"{message}\n", )
-    return {
+    output = {
         "message": message,
         "exception": ''.join(traceback.format_exception(tex.exc_type, tex, tex.exc_traceback))
     }
+    print(output)
```
[the remark this answers is not in the document]
Comment (cliffckerr) [reply covasim/webapp/cova_app.py:54]: Fixed

File: covasim/webapp/cova_app.py [hunk 6]
```diff
@@ -207,8 +160,71 @@ def get_gantt(intervention_pars=None, intervention_config=None):
 
     return response
 
+
+#%% Define the core API
+
+def parse_interventions(int_pars):
```
[the remark this answers is not in the document]
Comment (cliffckerr) [reply covasim/webapp/cova_app.py:166]: The interventions should mostly all have the same API so my concern with separate functions is that it would introduce a lot of duplication. We could/should probably even collapse the if statement a bit, pull parameters out to the front, etc.

File: covasim/webapp/index.html (hunk 2 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:67]: Fixed

File: covasim/webapp/index.html (hunk 3 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:328]: I tried `` but that didn't work...not sure what else needs to change

**Comment by cliffckerr** [comment]
@mfisher-idmod when uploading a CSV file of the right format (see tests/example_data.csv), the upload data button works for me. probably could have a better error message though. as for the reset button, can you make a ticket for that? reset only clears the BE parameters, not the FE ones, it seems

**Comment by cliffckerr** [comment]
@mfisher-idmod @lgeorge-idm i fixed this as well:
https://github.com/InstituteforDiseaseModeling/covasim/issues/173

**Comment by mfisher-idmod** [comment]
> @mfisher-idmod when uploading a CSV file of the right format (see tests/example_data.csv), the upload data button works for me. probably could have a better error message though. as for the reset button, can you make a ticket for that? reset only clears the BE parameters, not the FE ones, it seems

I was trying xlsx. I just tried csv with the Parameters tab from a previous sim run I got from downloading the xlsx from the web app, and that errors:
image

**Comment by ckerr-IDM** [comment]
yes it needs to be a different, as of yet undocumented, data format (see the csv file in tests for an example)...building the airplane while flying, fun times!

File: covasim/webapp/cova_app.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lgeorge-idm) [reply covasim/webapp/cova_app.py:166]: Duplication isn't the enemy. It can give you more flexibility in the future and make it more comprehensible to newcomers.
Given that the number and quantity of interventions is still in flux (assuming changes post-feedback from stakeholders), it would be premature to solidify the intervention api too much.

File: covasim/webapp/index.html [hunk 7]
```diff
@@ -36,199 +36,222 @@
...
   <div class="content">
 
-    <b-sidebar id="parameters" :visible="panel_open" no-header bg-variant="white" :width="panel_width+'vw'" shadow>
+    <b-sidebar id="parameters" :visible="panel_open" no-header class="sidebar" :width="panel_width+'vw'" shadow>
       <div class="px-3">
         <header class="d-flex justify-content-between">
-          <h4 class="my-2">Campaign Builder</h4>
-          <button class="close" @click="close_panel" aria-label="Close">
-            <span aria-hidden="true">&times;</span>
+          <h4 class="my-2">Simulation configuration</h4>
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:65]: Sentence casing is hard to parse when it's a header. I think you should revert all of these changes.

File: covasim/webapp/cova_app.css [hunk 8]
```diff
@@ -32,6 +48,18 @@ html, body {
     margin: 1em 0;
 }
 
+.sidebar {
+    background-color: #F8F9FA !important;
+}
+
+.parheading {
+    background-color:#000;
+    color: rgba(255,255,255,0.8); !important;
```
Comment (lgeorge-idm) [inline covasim/webapp/cova_app.css:57]: Invalid CSS - too many semicolons

File: covasim/webapp/cova_app.css [hunk 9]
```diff
@@ -32,6 +48,18 @@ html, body {
...
+}
+.parheading:hover {
+    color: rgba(255,255,255,1.0); !important;
```
Comment (lgeorge-idm) [inline covasim/webapp/cova_app.css:60]: Invalid CSS - too many semicolons

File: covasim/webapp/index.html [hunk 10]
```diff
@@ -36,199 +36,222 @@
...
-            </button>
-            <b-tooltip target="upload-parameters-button">
-              Upload parameters
-            </b-tooltip>
-          </div>
+      <div class="card mb-3">
+        <div class="card-header">
+          <button class="btn btn-block parheading" v-b-toggle.general>General options</button>
         </div>
+        <b-collapse visible id="general">
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:79]: All of the layout changes here (adding another collapse and  nesting the form rows) broke the sidebar maintaining the correct height on scroll after running the model. Please revert.

File: covasim/webapp/index.html [hunk 11]
```diff
@@ -330,9 +337,9 @@ <h4 class="my-2">Campaign Builder</h4>
           </svg>
           <span class="sr-only">View this project on GitHub</span>
         </b-nav-item>
-        <b-nav-item v-b-modal.licenses>License</b-nav-item>
-        <b-nav-item :href="app.org_url + '/terms'" target="_blank">Terms</b-nav-item>
-        <b-nav-item :href="app.org_url + '/privacy'" target="_blank">Privacy</b-nav-item>
+        <b-nav-item v-b-modal.licenses><span class="nav-text">License</span></b-nav-item>
+        <b-nav-item :href="app.org_url + '/terms'" target="_blank"><span class="nav-text">Terms</span></b-nav-item>
+        <b-nav-item :href="app.org_url + '/privacy'" target="_blank"><span class="nav-text">Privacy</span></b-nav-item>
       </b-navbar-nav>
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:343]: Overriding the nav-text class and adding the spans is unnecessary. Instead, just override the nav-link class.

File: covasim/webapp/index.html [hunk 12]
```diff
@@ -36,199 +36,222 @@
...
+      <p>Viral transmission occurs on a fixed contact network with undirected edges. Each day, infectious individuals expose susceptible individuals to possible infection. By default, the daily probability of infection, number of contacts, and mean duration of infectiousness roughly equates to R0 = 2.5 and a doubling time of 6-7 days. For more information, please visit the <a href="https://github.com/InstituteforDiseaseModeling/covasim" target="_blank">GitHub</a> repository.</p>
+      <p><b>Note:</b> Models are only as good as the values of the parameters put into them. This is a rapidly changing situation, and we will update content as relevant information is received.</p>
+      <p><b>Acknowledgements:</b> Covasim is currently under active development by IDM's COVID-19 Response Team. Model development is led by Daniel Klein, Cliff Kerr, Robyn Stuart, Romesh Abeysuriya, and Dina Mistry. UI design is led by Lauren George, William Chen, and Scott Ayers, inspired by a prototype from Brian Lovin. For the full list of contributors, please see our <a href="https://github.com/InstituteforDiseaseModeling/covasim/graphs/contributors" target="_blank">Contributors</a> page</a>.</p>
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:54]: Please use the {{app.github_url}} tokens.

File: covasim/webapp/index.html [hunk 13]
```diff
@@ -14,7 +14,7 @@
   <script src="https://polyfill.io/v3/polyfill.min.js?version=3.52.1&features=es2015%2CIntersectionObserver"></script>
   <script src="https://unpkg.com/vue@2.6.11/dist/vue.min.js"></script>
   <script src="https://unpkg.com/bootstrap-vue@2.11.0/dist/bootstrap-vue.min.js"></script>
-  <link href="https://fonts.googleapis.com/css?family=Roboto:wght@400;700&display=swap" rel="stylesheet">
+  <link href="https://fonts.googleapis.com/css?family=Quicksand:wght@400;700&display=swap" rel="stylesheet">
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:17]: Quicksand is hard to read on non-Ultra HD monitors. Please pick another font or choose a heavier default weight.

File: covasim/webapp/index.html [hunk 14]
```diff
@@ -36,199 +36,222 @@
...
-        <b-nav-item v-b-modal.about>About</b-nav-item>
+      <b-navbar-nav class="nav-head">
+        <b-nav-item :href="app.docs_url" target="_blank"><span class="nav-text">Documentation</span></b-nav-item>
+        <b-nav-item v-b-modal.about><span class="nav-text">About</span></b-nav-item>
       </b-navbar-nav>
       <b-navbar-nav class="ml-auto">
-        <b-nav-item :href="app.github_url" target="_blank">{{app.version}}</b-nav-item>
+        <b-nav-item :href="app.github_url" target="_blank"><span class="nav-text">{{app.version}}</span></b-nav-item>
       </b-navbar-nav>
     </b-navbar>
```
Comment (lgeorge-idm) [inline covasim/webapp/index.html:46]: Overriding the nav-text class and adding the spans is unnecessary. Instead, just override the nav-link class.

File: covasim/webapp/cova_app.css [hunk 15]
```diff
@@ -20,6 +21,21 @@ html, body {
...
+.navbar {
+    background-color: #111313 !important;
+}
+
+.nav-text {
+    color: rgba(255,255,255,0.8) !important;
+}
+.nav-text:hover {
+    color: rgba(255,255,255,1.0) !important;
+}
```
Comment (lgeorge-idm) [inline covasim/webapp/cova_app.css:33]: Overriding the nav-text class and adding the spans is unnecessary. Instead, just override the nav-link class.

File: covasim/webapp/cova_app.css [hunk 16]
```diff
@@ -32,6 +48,18 @@ html, body {
     margin: 1em 0;
 }
 
+.sidebar {
+    background-color: #F8F9FA !important;
+}
+
+.parheading {
```
Comment (lgeorge-idm) [inline covasim/webapp/cova_app.css:55]: Would be better to override the card-header class.

**Comment by lgeorge-idm** [comment]
Example of the sidebar height issues:
image

File: covasim/webapp/index.html (hunk 3 above)
Comment (lgeorge-idm) [reply covasim/webapp/index.html:327]: That's because we're using Bootstrap 4 (this is why I pinned the versions, it makes it easier to look up the docs and such!). The class is 'img-fluid'. You should also add in cova_app.css a max-height for the footer images of 2em and make sure the text aligns properly.

File: covasim/webapp/cova_app.js [hunk 17]
```diff
@@ -156,7 +156,7 @@ var vm = new Vue({
                 local_path: null,
                 server_path: null
             },
-            intervention_pars: {},
+            int_pars: {},
```
[the remark this answers is not in the document]
Comment (cliffckerr) [reply covasim/webapp/cova_app.js:159]: Resolved in next comment I think?

File: covasim/webapp/cova_app.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cliffckerr) [reply covasim/webapp/cova_app.py:166]: It's definitely a balance, and if we end up with something significantly more complicated, we might want to refactor. Since I'll probably be the one working the most on this part of the code for the time being, and since this will be easiest the way things currently are, I'd prefer to keep it as-is for now

File: covasim/webapp/index.html (hunk 3 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:327]: I agree this would be better but I won't have time to fix straight away -- I'll make an issue

File: covasim/webapp/index.html (hunk 7 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:65]: So I was wondering why I'm averse to title case. TIL, it's an Americanism:

https://www.herosmyth.com/article/headline-capitalization-ux-title-case-vs-sentence-case

"If your audience is only in the US, then reference other factors to determine which case to implement in your headlines.

Otherwise, if you’re writing internationally-targeted content, think global—use sentence case to cater to more audiences rather than fewer."

So strong vote from this Aussie to keep sentence case ;)

File: covasim/webapp/cova_app.css (hunk 8 above)
Comment (cliffckerr) [reply covasim/webapp/cova_app.css:57]: Stupid webstorm lack-of-linting. Fixed

File: covasim/webapp/cova_app.css (hunk 9 above)
Comment (cliffckerr) [reply covasim/webapp/cova_app.css:60]: Fixed

File: covasim/webapp/index.html (hunk 10 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:79]: I think we need to keep the collapsible "General" options, but have added it to the refactor: https://github.com/InstituteforDiseaseModeling/covasim/issues/180

File: covasim/webapp/index.html (hunk 11 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:343]: Agree, added to https://github.com/InstituteforDiseaseModeling/covasim/issues/180

File: covasim/webapp/index.html (hunk 12 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:54]: This isn't working for some reason, will add to issue: 
image

File: covasim/webapp/index.html (hunk 7 above)
Comment (lgeorge-idm) [reply covasim/webapp/index.html:65]: Fair enough. I'm an American working in an American org. My bias is showing :).

File: covasim/webapp/index.html (hunk 10 above)
Comment (lgeorge-idm) [reply covasim/webapp/index.html:79]: I don't think we need to collapse the general options. They're general because we decided in that meeting on the 7th that those were the most common options. Why would you hide them?

File: covasim/webapp/index.html (hunk 13 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:17]: I changed to 
```
<link href="https://fonts.googleapis.com/css?family=Quicksand:wght@600;900&display=swap" rel="stylesheet">
```
(is that right?) Can you share a screenshot of the low-res readability issue? It looked fine on my phone

File: covasim/webapp/index.html (hunk 14 above)
Comment (cliffckerr) [reply covasim/webapp/index.html:46]: Added to issue

File: covasim/webapp/cova_app.css (hunk 15 above)
Comment (cliffckerr) [reply covasim/webapp/cova_app.css:33]: Added to issue

File: covasim/webapp/cova_app.css (hunk 16 above)
Comment (cliffckerr) [reply covasim/webapp/cova_app.css:55]: Added to issue

File: covasim/webapp/index.html (hunk 13 above)
Comment (lgeorge-idm) [reply covasim/webapp/index.html:17]: I can't share a screenshot as your monitor is the issue. I would have to take a photo of the monitor itself. 
I was looking at it on a 1080p monitor on windows (the most common resolution on the internet). Thin fonts just don't render as well and most phones nowadays have higher res screens than that.

File: covasim/webapp/index.html (hunk 10 above)
Comment (ckerr-IDM) [reply covasim/webapp/index.html:79]: If you are working with a lot of interventions, it can get a bit cramped otherwise. I was getting annoyed at not being able to collapse to save screen real estate

File: covasim/webapp/index.html (hunk 13 above)
Comment (ckerr-IDM) [reply covasim/webapp/index.html]: This is maybe overkill but the useragent has screen resolution info doesn't it? Could pick the font responsively

Comment (ckerr-IDM) [reply covasim/webapp/index.html]: Or if Windows is detected, set the font to Courier New with 8-bit colors ;)

Comment (lgeorge-idm) [reply covasim/webapp/index.html]: That is more work than is necessary. The updated URL is correct, but you'll need to override the body font-weight to 600 too.

Comment (lgeorge-idm) [reply covasim/webapp/index.html]: Picking a font based on useragent is a nightmare waiting to happen because useragents lie all the time.

Comment (ckerr-IDM) [reply covasim/webapp/index.html]: that's true, if you're lying about your screen resolution then maybe you deserve what's coming ;) tried heavier weight, it's...not great: 
image
suggest leaving for now. we do have a user group who we could ask, and can certainly revisit if it comes up

Comment (mfisher-idmod) [reply covasim/webapp/index.html:17]: Added to issue #180
